### ISSUE TITLE
[GLES2, Batching] If a TextureRect uses stretch_mode Tile it breaks the texture on Sprites using repeat

### ISSUE THREAD

**Issue by noidexe** [godotengine/godot#39766]
**Godot version:**
Confirmed in 3.2.2 rc2 and 3

**OS/device including version:**
Windows, Android, HTML5 
Happens only when using GLES2 with batching enabled

**Issue description:**
 - You the same texture assigned to a Sprite and a TexutureRect. 
 - The texture was imported with repeat enabled. 
 - The sprite uses region to tile the texture.
 -  he TextureRect uses stretch_mode: tile to tile the texture. 
When the TextureRect appears on screen for the first time it corrupts the texture whenever it is displayed on any sprite. 
Disabling batching solves the issue. Enabling batching and single rect fallback breaks the texture on both the sprite and the texture rect. 

Possibly related: Tiling doesn't work with 9 patch rects under the same conditions(gles2, batching)

**Steps to reproduce:**
- open http://crimsonghostgames.com/tilingbug/ on a browser
- click start
- notice dark rock background (Sprite)
- click home icon on top left
- click on credits
- notice dark rock background (TextureRect)
- click home icon on top right
- click on start again
Expected:
- The background should look exactly like the first time
Actual:
- The background texture is displayed incorrectly (looks like it's just repeating the border pixels)

**Minimal reproduction project:**
Batching enabled, bug present: http://crimsonghostgames.com/tilingbug/
Batching disabled, bug not present: http://crimsonghostgames.com/pichon/

**Comment by clayjohn** [comment]
Cc @lawnjelly

**Comment by lawnjelly** [comment]
Looks like we might be missing a texture clamp mode. I'll probably need a minimum reproduction project though in order to solve this (i.e. project, not exported project).

**Comment by lawnjelly** [comment]
This is actually very interesting bug. I've worked out why it happens (in the batching, I'm not checking the texture repeat flag, only the rect repeat flag). However, it turns out there may be a bug in the legacy renderer as well, it is just not exhibiting.

A single sprite with a repeating texture using a region is NOT set to repeat mode by the legacy code. It only works because it happens to be in repeat mode (i.e. by accident). I.e. the existing behaviour is, in some senses, incorrect.

The batching code is not completely correct, but according to the existing logic it is correct not to wrap the texture. However I suspect that the existing logic is wrong.

This is the logic in the legacy renderer:
```
if (can_tile && r->flags & CANVAS_RECT_TILE && !(tex->flags & VS::TEXTURE_FLAG_REPEAT)) {
glTexParameterf(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, GL_REPEAT);
glTexParameterf(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, GL_REPEAT);
untile = true;
}
```
This appears to only set wrapping if both the rect flag is set to CANVAS_RECT_TILE and the tex flags are NOT set to VS::TEXTURE_FLAG_REPEAT. This I'm presuming is a bug, and the NOT is incorrect.

Two behaviours would seem more reasonable. Either:
1) Wrap when EITHER the texture wrap flag is set or the rect wrap flag
2) Wrap only when both the rect wrap flag AND the texture wrap flag are set.

I may have to consult with reduz on what the intended behaviour is. Regardless we will need something consistent with the existing behaviour.

**Comment by akien-mga** [comment]
Fixed by #39774.
